### DISCUSSION TITLE
How to structure Flux Kustomizations for single repository?

### DISCUSSION THREAD

**Discussion by mloskot** [fluxcd/flux2#3801] [Q&A] [answered]
Are there any best practices or recommendations about how to 'structure' Flux Kustomizations defined for a single repository?

As far as I understand, there are options, for example:
- one Kustomization for everything
- one Kustomization for core/system resources and another separate one for end-user apps?
- one Kustomization per tightly related group of resources (e.g. per environment, per team, per namespace)

Does it even matter?

Additionally, does it matter or affect Flux operations in any way if two Kustomizations have overlapping path -s like these two below?

![image](https://user-images.githubusercontent.com/80741/232230624-a4992be2-7772-497e-8b78-ae15aa8360ad.png)

I looked at the FAQ, the End-to-End and [Kustomization reocmmended settings](https://fluxcd.io/flux/components/kustomize/kustomization/#recommended-settings) pages but those don't seem to discuss that. I guess, I may be not the only Flux newbie wondering about it. So, I have also [asked on #flux channel on Slack](https://cloud-native.slack.com/archives/CLAJ40HV3/p1681413332196719), but the thread has not received too much feedback.

**Comment by stefanprodan** [comment, marked as answer]
> Additionally, does it matter or affect Flux operations in any way if two Kustomizations have overlapping path

Yes, the reconcilers will fight over the resources, you should make sure this doesn't happen. It's also easy to detect this, you'll get huge spam on Slack or any other notification provider you've configured. Please use the https://github.com/fluxcd/flux2-kustomize-helm-example as a starting point to avoid such situations. 

> Are there any best practices or recommendations about how to 'structure' Flux Kustomizations defined for a single repository?

At minimum you should have two Kustomizations besides the `flux-system` one. One for cluster addons and one for apps. The Flux Kustomization is intended to manage a single app, doing so, you have better control over that app rollout, during an incident you can suspend/resume it without affecting all others. But if the apps are deployed via HelmReleases, then it's perfectly fine to have a single Kustomization applying them all, since you have control over each app using the HelmRelease object. This what the example repo shows.

**Reply by mloskot** [reply]
@stefanprodan Thank you very much for such an excellent answer - exactly what I was missing and looking for.
Additionally, you've helped me understand why I'm spammed with notifications 😊

**Reply by mloskot** [reply]
For benefit of readers of this thread, take @stefanprodan 's warning above about avoiding overlapping `path`-s seriously:

>  reconcilers will fight over the resource

Here is an interesting case of using External Secrets Operator feature of `Password` generator and `ExternalSecret` in one of my clusters which apparently had not been functioning correctly due to the multiple Flux Kustomizations with overlapping paths:
- https://github.com/external-secrets/external-secrets/discussions/2402#discussioncomment-6163733

Please, notice the disclaimer I posted in that ESO thread:

> I can't explain why overlapping paths would affect Flux such that ESO re-runs Password generator referred by my API key secrets, but that is what I am observing - getting rid of overlapping paths fixes the problem.

So, my conclusions may change as I'm still learning :-)

**Reply by reachjyo** [reply]
I have the following use case so checking you can help us out ?
I have namespace **temp** and there are couple of services deployed . I have added FLUXCD notification to receive helm chart deployment notifications for those services  in **temp** channel and Iam receiving the notifications as expected. Iam going to add couple  more services  to the existing K8s cluster but plan to receive the deployment notifications in a different channel ie **temp-new**  channel ? My understanding is flux CD configuration is done 1 time at the cluster level and I am not sure if its allowed to update the configuration to receive notifications for new services in a different channel from the same cluster ? Appreciate if anyone can shed some light on this?
